Make the resampler use timedelta for periods

leandro-lucarella-frequenz opened the pull request:
As a step towards [using proper units], now the resampler uses `timedelta` for specifying periods. It also includes a couple of cleanup commits.

- Move private method to the end of the class
- Remove unused timer
- Make the resampling period a `timedelta`
- Make source properties' sampling period a `timedelta`

[using proper units]: https://github.com/frequenz-floss/frequenz-sdk-python/discussions/6

llucax committed:
Move private method to the end of the class

This is only to make it easier to read the class, having the public
methods first and then the private ones.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

llucax committed:
Remove unused timer

Now the `_wait_for_next_resampling_period()` method is used to trigger
the next resampling.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

llucax committed:
Make the resampling period a `timedelta`

The `ResamplerConfig` now takes the resampling period as a `timedelta`.

The configuration was renamed from `resampling_period_s` to
`resampling_period` accordingly.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

llucax committed:
Make source properties' sampling period a `timedelta`

The `SourceProperties` of the resampler now uses a `timedelta`
for the input sampling period. The attribute was renamed from
`sampling_period_s` to `sampling_period` accordingly.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/_resampling.py:
```diff
@@ -657,9 +657,12 @@ def resample(self, timestamp: datetime) -> Sample:
+                props.sampling_period.total_seconds(),
+            )
+            if props.sampling_period is not None
+            else conf.resampling_period.total_seconds()
```
I think period should be also a `timedelta` type and the 3 calls to `.total_seconds()` above can be avoided if I am not missing anything. Then call `period.total_seconds()`  when it is needed instead

leandro-lucarella-frequenz replied:
Yeah, good catch, it does look simpler.

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/_resampling.py:
```diff
@@ -583,26 +581,28 @@ def _update_buffer_len(self) -> bool:
+                config.resampling_period.total_seconds()
+                / input_sampling_period.total_seconds()
                 * config.max_data_age_in_periods
             )
```
not entirely related to this PR, but I'd like to mention it anyway :) I think this is not a good practice in the sense that here `new_buffer_len` is defined dynamically as a `float` where in other places above it is an int. So we should probably use `math.ceil()` here instead

leandro-lucarella-frequenz replied:
Yeah, the `ceil` is done afterwards:

```py
        else:
            new_buffer_len = (
                config.resampling_period.total_seconds()
                / input_sampling_period.total_seconds()
                * config.max_data_age_in_periods
            )

        new_buffer_len = max(1, math.ceil(new_buffer_len))
#                               ^^^^^^^^^^^^^^^^^^^^^^^^^
```

It is weird though that the `new_buffer_len` is a `float`, so I will add a commit to improve the typing.

daniel-zullo-frequenz reviewed (dismissed):
Just an optional comment to check for, LGTM

llucax committed:
Improve typing while calculating the new buffer len

`new_buffer_len` has `float` type, which is weird, even if it is handled
properly and converted when assigning to the buffer's `maxlen`. This
commit makes `new_buffer_len` always an `int` to make the intentions
more clear.

Signed-off-by: Leandro Lucarella <leandro.lucarella@frequenz.com>

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/_resampling.py:
```diff
@@ -583,28 +581,30 @@ def _update_buffer_len(self) -> bool:
             )
+        )
 
         new_buffer_len = max(1, math.ceil(new_buffer_len))
```
then `math.ceil()` might not be needed here based on the latest changes if I'm not mistaken

leandro-lucarella-frequenz commented:
Rebased and updated to address the issues commented above. I enabled auto-merge, so if anyone approves it will be merged. Since it was pre-approved, if nobody approves in a while I will force-merge anyway :)

daniel-zullo-frequenz commented:
@leandro-lucarella-frequenz have you seen this comment https://github.com/frequenz-floss/frequenz-sdk-python/pull/334#discussion_r1169864522 ?

mathias-baumann-frequenz commented:
Needs rebase

leandro-lucarella-frequenz commented:
> @leandro-lucarella-frequenz have you seen this comment [#334 (comment)](https://github.com/frequenz-floss/frequenz-sdk-python/pull/334#discussion_r1169864522) ?

Yes, good catch, thanks and thanks for disabling auto-merge! Updated and re-enabled auto-merge.
